iroh-gossip/src/net.rs

```diff
@@ -0,0 +1,819 @@
+                        self.pending_sends.entry(peer_id).or_default().push(message);
+                    }
+                }
+                OutEvent::EmitEvent(topic_id, event) => {
+                    if let Some(sender) = self.subscribers_all.as_mut() {
+                        if let Err(_event) = sender.send((topic_id, event.clone())) {
+                            self.subscribers_all = None;
+                        }
+                    }
+                    if let Some(sender) = self.subscribers_topic.get(&topic_id) {
+                        // Only error case is that all [broadcast::Receivers] have been dropped.
+                        // If so, remove the sender as well.
+                        if let Err(_event) = sender.send(event) {
+                            self.subscribers_topic.remove(&topic_id);
+                        }
+                    }
+                }
+                OutEvent::ScheduleTimer(delay, timer) => {
+                    self.timers.insert(now + delay, timer);
+                }
+                OutEvent::DisconnectPeer(peer) => {
+                    if let Some(conn) = self.conns.remove(&peer) {
+                        conn.close(0u8.into(), b"close from disconnect");
+                    }
+                    self.conn_send_tx.remove(&peer);
+                    self.pending_sends.remove(&peer);
+                    self.dialer.abort_dial(&peer);
+                }
+                OutEvent::PeerData(peer, data) => match postcard::from_bytes::<IrohInfo>(&data) {
+                    Err(err) => warn!("Failed to decode PeerData from {peer}: {err}"),
+                    Ok(info) => {
+                        debug!("add known addrs for {peer}: {info:?}...");
+                        self.endpoint
+                            .add_known_addrs(peer, info.derp_region, &info.addrs)
+                            .await?;
+                    }
+                },
+            }
+        }
+        Ok(())
+    }
+
+    fn subscribe_all(&mut self) -> broadcast::Receiver<(TopicId, Event)> {
+        if let Some(tx) = self.subscribers_all.as_mut() {
+            tx.subscribe()
+        } else {
+            let (tx, rx) = broadcast::channel(SUBSCRIBE_ALL_CAP);
+            self.subscribers_all = Some(tx);
+            rx
+        }
+    }
+
+    fn subscribe(&mut self, topic_id: TopicId) -> broadcast::Receiver<Event> {
+        if let Some(tx) = self.subscribers_topic.get(&topic_id) {
+            tx.subscribe()
+        } else {
+            let (tx, rx) = broadcast::channel(SUBSCRIBE_TOPIC_CAP);
+            self.subscribers_topic.insert(topic_id, tx);
+            rx
+        }
+    }
+}
+
+async fn wait_for_neighbor_up(mut sub: broadcast::Receiver<Event>) -> anyhow::Result<()> {
+    loop {
+        match sub.recv().await {
+            Ok(Event::NeighborUp(_neighbor)) => break Ok(()),
+            Ok(_) | Err(broadcast::error::RecvError::Lagged(_)) => {}
```

Comment:
self-review: if the channel lags, this could lead to an `Event::NeighborUp` to be missed, causing `GossipHandle::join` to never complete. Don't think this can be a problem in practice, because the loop does nothing for messages, so it should just never occur that there's a `RecvError::Lagged` (events would have to be produced faster than a loop that does nothing can consume).